Shared/Tests.Sync.Shared/MergeByPKTests.cs

```diff
@@ -0,0 +1,122 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using IntegrationTests;
+using NUnit.Framework;
+using Realms;
+using Realms.Sync;
+
+namespace Tests.Sync
+{
+    [TestFixture, Preserve(AllMembers = true)]
+    public class MergeByPKTests
+    {
+        [NUnit.Framework.Explicit]
+        [TestCaseSource(nameof(MergeTestCases))]
+        public async void WhenObjectHasPK_ShouldMergeCorrectly(Type objectType, object pkValue, Func<dynamic, bool> pkValueChecker)
+        {
+            User.ConfigurePersistence(UserPersistenceMode.NotEncrypted);
+
+            var pkProperty = objectType.GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                                       .Single(p => p.GetCustomAttribute<PrimaryKeyAttribute>() != null);
+
+            for (var i = 0; i < 5; i++)
+            {
+                var instance = (RealmObject)Activator.CreateInstance(objectType);
+
+                pkProperty.SetValue(instance, (dynamic)pkValue);
+
+                using (var realm = await GetSyncedRealm(objectType))
+                {
+                    try
+                    {
+                        realm.Write(() => realm.Add(instance));
+                    }
+                    catch (RealmDuplicatePrimaryKeyValueException)
+                    {
+                        // Sync went through too quickly (that's why we do 5 attempts
+                    }
+
+                    await Task.Delay(500); // A little bit of time for sync
+                }
+
+                User.Current.LogOut();
```

Comment:
I wanted to simulate a scenario where a user logs in for the first time and opens a realm. I guess that's too much - I'll simplify it.